lightning-custom-message/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+//! }
+//!
+//! #[derive(Debug)]
+//! pub struct Bar;
+//!
+//! macro_rules! bar_type_id {
+//!     () => { 32769 }
+//! }
+//!
+//! impl wire::Type for Bar {
+//!     fn type_id(&self) -> u16 { bar_type_id!() }
+//! }
+//! impl Writeable for Bar {
+//!     // ...
+//! #     fn write<W: Writer>(&self, _: &mut W) -> Result<(), io::Error> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//!
+//! pub struct BarHandler;
+//!
+//! impl CustomMessageReader for BarHandler {
+//!     // ...
+//! #     type CustomMessage = Bar;
+//! #     fn read<R: io::Read>(
+//! #         &self, _message_type: u16, _buffer: &mut R
+//! #     ) -> Result<Option<Self::CustomMessage>, DecodeError> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//! impl CustomMessageHandler for BarHandler {
+//!     // ...
+//! #     fn handle_custom_message(
+//! #         &self, _msg: Self::CustomMessage, _sender_node_id: &PublicKey
+//! #     ) -> Result<(), LightningError> {
+//! #         unimplemented!()
+//! #     }
+//! #     fn get_and_clear_pending_msg(&self) -> Vec<(PublicKey, Self::CustomMessage)> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//!
+//! #[derive(Debug)]
+//! pub struct Baz;
+//!
+//! macro_rules! baz_type_id {
+//!     () => { 32770 }
+//! }
+//!
+//! impl wire::Type for Baz {
+//!     fn type_id(&self) -> u16 { baz_type_id!() }
+//! }
+//! impl Writeable for Baz {
+//!     // ...
+//! #     fn write<W: Writer>(&self, _: &mut W) -> Result<(), io::Error> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//!
+//! pub struct BazHandler;
+//!
+//! impl CustomMessageReader for BazHandler {
+//!     // ...
+//! #     type CustomMessage = Baz;
+//! #     fn read<R: io::Read>(
+//! #         &self, _message_type: u16, _buffer: &mut R
+//! #     ) -> Result<Option<Self::CustomMessage>, DecodeError> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//! impl CustomMessageHandler for BazHandler {
+//!     // ...
+//! #     fn handle_custom_message(
+//! #         &self, _msg: Self::CustomMessage, _sender_node_id: &PublicKey
+//! #     ) -> Result<(), LightningError> {
+//! #         unimplemented!()
+//! #     }
+//! #     fn get_and_clear_pending_msg(&self) -> Vec<(PublicKey, Self::CustomMessage)> {
+//! #         unimplemented!()
+//! #     }
+//! }
+//!
+//! # fn main() {
+//! composite_custom_message_handler!(
+//!     pub struct FooBarHandler {
+//!         foo: FooHandler,
+//!         bar: BarHandler,
+//!     }
+//!
+//!     pub enum FooBarMessage {
+//!         Foo(foo_type_id!()),
+//!         Bar(bar_type_id!()),
+//!     }
+//! );
+//!
+//! macro_rules! foo_bar_type_ids {
+//!     () => { foo_type_id!() | bar_type_id!() }
+//! }
+//!
+//! composite_custom_message_handler!(
+//!     pub struct FooBarBazHandler {
+//!         foo_bar: FooBarHandler,
+//!         baz: BazHandler,
+//!     }
+//!
+//!     pub enum FooBarBazMessage {
+//!         FooBar(foo_bar_type_ids!()),
+//!         Baz(baz_type_id!()),
+//!     }
+//! );
```

Comment:
> Is there a reason someone would opt for using the macro multiple times like this as opposed to just listing each message + handler in one? i.e.
> 
> ```
> composite_custom_message_handler!(
>   pub struct FooBarBazHandler {
>     foo: FooHandler,
>     bar: BarHandler,
>     baz: BazHandler,
>   }
>   ...
> }
> ```

That is also possible, though the example is meant to demonstrate arbitrary composition. This may be useful when `FooBarHandler` is defined in one crate and you want to use it along with `BazHandler` in your own crate. I've added some comments to the example to clarify this use case.

> Also is there significance to the `foo_bar_type_ids` being the `|` of the two type ids or was that mostly arbitrary? Otherwise, I thought the `FooBar` example was very helpful/straightforward!

Yes, the macro matcher `$pattern` uses a `pat` specifier, meaning it can be anything rust considers a pattern, including the composition of patterns using `|`.

https://doc.rust-lang.org/reference/patterns.html

Someone using the macro could use a range of type ids instead, e.g., `32768.. =32769`. Though the `|` approach is preferred for the reason given in the macro's docs.

Reply:
> That is also possible, though the example is meant to demonstrate arbitrary composition. This may be useful when FooBarHandler is defined in one crate and you want to use it along with BazHandler in your own crate. I've added some comments to the example to clarify this use case.

Ah makes sense!

> Yes, the macro matcher `$pattern` uses a `pat` specifier, meaning it can be anything rust considers a pattern, including the composition of patterns using `|`.

Ohh I see I didn't realize it was being used as a pattern I thought it doing a bitwise OR of the two type ids, that's cool.